design/web-authn-authenticator.md

```diff
@@ -0,0 +1,425 @@
+  * If matches, register a public key credential based on this information.
+  * If not, returns an error telling a user .
+
+This AAGUID list works as white list.
+
+
+#### by Metadata acquired from extrenal sources
+
+At the first step of implementation, it is out of scope. After realizing support authenticator metadata aquisition from external sources (e.g. FIDO Alliance Metadata Services), I'll consider filtering users' authenticators based on such the metadata.
+
+### Open Issues
+
+User Interface, Logging, Error Handling need to be considered.
+
+## Authentication
+***
+
+### Definition
+
+"Authentication" here means that a user use their public key credential as credential of their user authentication.
+
+### How to Realize
+
+It is realized by an authenticator provider called "WebAuthn Authenticator". It conforms to the following acceptance criteria in [KEYCLOAK-9360](https://issues.jboss.org/browse/KEYCLOAK-9360) : Two factor authentication with W3C Web Authentication:
+
+>    * Admin manually registers Web Authentication authenticator to replace OTP authenticator
+
+It is an authenticator provider. Therefore, by adding this authenticator provider after existing username authenticator provider, it conforms to the following acceptance criteria in [KEYCLOAK-9360](https://issues.jboss.org/browse/KEYCLOAK-9360) : Two factor authentication with W3C Web Authentication:
+
+>    * When user logs in again after registering the security key the user should be prompted to tap the security key after entering username and password
+
+Also, to make the browser execute WebAuthenticatin API, javascript and freemarker template are required.
+
+### Configuration
+
+Authentication can be configured by "WebAuthn Authenticator" provider config. 
+
+Basically, this design document follows [WebAuthn API Specification](https://www.w3.org/TR/webauthn/#assertion-options) for configuration items and their values.
+
+#### Configuration Scope
+
+This configuration is applied to all users using this provider. It suffice to say that this configuration is adopted per realm.
+
+#### Configuration Item - User Verification Requirement
+
+  - Have the switch (ON/OFF) to specify whether this configuration is used or not. Its default setting is OFF.
+  - Only single item can be selected.
+  - Supported user verification options are the following : {required, preferred, discouraged}
+  - Default setting is : preferred
+
+*Notes:*
+
+* This setting is optional for WebAuthn API (`navigator.credentials.get()`). Therefore, if the switch if OFF, the executor of this WebAuthn API does not filter out authenticators by their user verification capability.
+
+### Specifying Public Key Credentials
+
+User can select which their public key credentials are used for their authentication.
+
+On user authentication:
+
+  * Keycloak shows the list of public key credentials with its metadata that the user has registered.

```

Comment:
Question: At stage 1, we will have just one public key per user. So would it makes sense, that if user has just one public key (which will be always at phase 1), the authenticator doesn't need to ask user, but automatically select that one public key? I guess it is not the case and it will be always good practice to ask user even if he has just one key, so that he confirms that he wants to use it?

I am not requesting to change anything, just asking to clarify this...

Reply:
At stage 1, I will implement the features about Public Key Credential (e.g. credential selection on authentication)  the same as the later stage when multiple credentials are supported in order that make it easy to work on lager stages (e.g. multiple credentials per user supported ). Therefore, I will create the public key credential selection page on authentication time even if only single public key credential is supported at the stage 1. 

It seems to be quite odd for users that the features supposing multiple credentials are implemented at state 1. But as @stianst commented on this pull request, this single public key credential is tentative and not officially supported later on. Therefore, I think this implementation strategy at state 1 might be acceptable. WDYT?

If so, is it better to mention this in the design document?

Reply:
Thanks for the clarification. I agree to keep as it is and have the selection screen present even if there is just one key. Not sure whether to add a NOTE to this design, that selection will be useful especially in later stages or whether to keep design as is. I am fine with both options and leave to you.

Reply:
I'll add this notification on Public Key Credentials section.